File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/Lift.java
package org.firstinspires.ftc.teamcode.subsystems;

import static org.firstinspires.ftc.teamcode.core.Constants.*;

import com.acmerobotics.dashboard.config.Config;
import com.arcrobotics.ftclib.command.SubsystemBase;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.arcrobotics.ftclib.controller.PIDFController;

import org.firstinspires.ftc.teamcode.core.MyRobot;
import org.firstinspires.ftc.teamcode.state.SubsystemStates.OuttakeState;

@Config
public class Lift extends SubsystemBase {

    private DcMotorEx motor1;
    private DcMotorEx motor2;
    private MyRobot robot;

    double p = 1;
    double i = 0;
    double d = 0.1;
    double f = 0;

    PIDFController pidf = new PIDFController(p, i, d, f);

    public Lift(MyRobot robot) {
        this.motor1 = robot.vslide1;
        this.motor2 = robot.vslide2;
        this.robot = robot;

        motor1.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        motor2.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);

        //encoder is used to be able to set velocity of motors
        //we should use a velocity pidf controller because we need precise control with low overshoot, and we dont need time to be super optimised
        motor1.setMode(DcMotor.RunMode.RUN_USING_ENCODER);
        motor2.setMode(DcMotor.RunMode.RUN_USING_ENCODER);

        pidf.setTolerance(LIFT_TOLERANCE);
    }

    public void liftUp() {
        pidf.setSetPoint(LIFT_UP);
        wait_for_lift();
    }

    public void liftDown() {
        pidf.setSetPoint(LIFT_DOWN);
        wait_for_lift();
    }

    public void goToPos(int pos) {
        pidf.setSetPoint(pos);
        wait_for_lift();
    }

    private void wait_for_lift() {
        while(!pidf.atSetPoint()) {
            double output = pidf.calculate(motor1.getCurrentPosition());
            motor1.setVelocity(output);
            motor2.setVelocity(output);
        }

        motor1.setVelocity(0);
        motor2.setVelocity(0);
    }


}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/Lift.kt
package org.firstinspires.ftc.teamcode.subsystems

import android.hardware.SensorPrivacyManager.Sensors
import com.acmerobotics.dashboard.config.Config
import com.arcrobotics.ftclib.command.SubsystemBase
import com.arcrobotics.ftclib.controller.PIDFController
import com.qualcomm.robotcore.hardware.DcMotor
import org.firstinspires.ftc.teamcode.core.Constants.*
import org.firstinspires.ftc.teamcode.core.Robot
import kotlin.math.abs
import kotlin.math.min


@Config
class Lift(val robot: Robot) : SubsystemBase() {
    val right = robot.vslider
    val left = robot.vslidel
    val touch = robot.vSlideTouch

    val p = 6.0
    val i = 0.0
    val d = 0.02
    val f = 0.0

    var doIStopPID = true
    var doIStopBrake = true

    val pidf: PIDFController = PIDFController(p, i, d, f)

    init {
        right.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        left.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER

        right.mode = DcMotor.RunMode.RUN_USING_ENCODER
        left.mode = DcMotor.RunMode.RUN_USING_ENCODER

        right.targetPositionTolerance = LIFT_TOLERANCE.toInt()
        left.targetPositionTolerance = LIFT_TOLERANCE.toInt()

        pidf.setTolerance(LIFT_TOLERANCE)
    }

    fun customSetPoint(setPoint: Double) { pidf.setPoint = setPoint }
    fun contractionSetPoint() { customSetPoint(LIFT_DOWN) }
    fun extensionSetPoint() { customSetPoint(LIFT_UP) }
    fun specWallSetPoint() { customSetPoint(LIFT_SPECIMEN_WALL)}
    fun specBarSetPoint() { customSetPoint(LIFT_SPECIMEN_BAR)}

    fun setVelocity(vel: Double) {
        right.velocity = vel
        left.velocity = -vel
    }

    fun updateVelocity() {
        setVelocity(pidf.calculate(right.currentPosition.toDouble()))
    }

    fun setModes(mode: DcMotor.RunMode) {
        right.mode = mode
        left.mode = mode
    }

    fun setPowers(power: Double) {
        left.power = power
        right.power = -power
    }

    fun atSetPoint() = pidf.atSetPoint()

    fun brake(power: Double) {
//        left.targetPosition = left.currentPosition
//        right.targetPosition = right.currentPosition
//        setModes(DcMotor.RunMode.RUN_TO_POSITION)
//        setPowers(power)
        //setVelocity(0.0)

        var error = abs(LIFT_UP - right.currentPosition)
        var final = 0.0

        if (error > LIFT_TOLERANCE) {
            var kp = 0.00001
            final = min(abs(power), abs(error*kp))
        }

        setPowers(final)
    }

    fun brake(power: Double, offset: Int) {
//        left.targetPosition = left.currentPosition + offset
//        right.targetPosition = right.currentPosition + offset
//        setModes(DcMotor.RunMode.RUN_TO_POSITION)
//        setPowers(power)
        customSetPoint(right.currentPosition.toDouble())
    }

    fun convertToInt(value: Any): Int {
        if (value is Int) return value
        else if (value is Double) return value.toInt()
        else {throw ClassCastException()}
    }
}